src/strategies/flex_window.py

```diff
@@ -291,37 +457,401 @@ def distribute_peak_shaving(self, timesteps):
+                if needed_power < 0:
+                    discharge = 0
+                else:
+                    discharge = battery.unload(
+                        self.interval, (needed_power/len(batteries)),
+                        target_soc=self.DISCHARGE_LIMIT
+                    )["avg_power"]
+                discharging_stations.append(b_id)
+                gc.add_load(b_id, -discharge)
+                timesteps[0]["total_load"] -= discharge
+
+    def distribute_peak_shaving_v2g(self, timesteps, commands):
+
+        """
+        # stationary batteries
+        for bid, battery in self.world_state.batteries.items():
+            if self.LOAD_STRAT == 'balanced':
+                # find minimum power to charge battery during windows
+        """
+        if not commands:
+            commands = {}
+        gc = list(self.world_state.grid_connectors.values())[0]
+        min_soc = 0.8
+        # get all vehicles that are connected in this step and order vehicles
+        vehicles = sorted(
+            [
+                v
+                for v in self.world_state.vehicles.values()
+                if v.connected_charging_station is not None
+            ],
+            key=self.sort_key,
+        )
+        cur_window = timesteps[0]["window"]
+
+        sim_vehicles = deepcopy(vehicles)
+        cur_time = self.current_time - self.interval
+        cur_vehicles = sim_vehicles
+
+        for i, v in enumerate(vehicles):
+            cs_id = v.connected_charging_station
+            cs = self.world_state.charging_stations[cs_id]
+
+            cur_vehicles = [
+                v
+                for v in cur_vehicles
+                if (v.estimated_time_of_departure > cur_time)
+                and (v.battery.soc < v.desired_soc)
+                and (v.vehicle_type.v2g)
+            ]
+            # check if cehicles can be loaded until desired_soc in connected timesteps
+
+            old_soc = v.battery.soc
+            connected_timesteps = []
+            window = cur_window
+            window_change = 0
+            # get connected timesteps and count number of window changes
+            if cur_vehicles:
+                index = 0
+                for ts_info in timesteps:
+                    index += 1
+                    cur_time += self.interval
+                    if ts_info["window"] != window:
+                        window_change += 1
+                        window = ts_info["window"]
+                    if v.estimated_time_of_departure < cur_time:
+                        break
+                    connected_timesteps.append(ts_info)
+            # always load if cur_window is True
+            if cur_window:
+                mode = "charge"
+            else:
+                mode = "discharge"
+                # check if vehicle ends up with desired soc, adjust min_soc accordingly
+                if window_change >= 1:
+                    min_soc = 0
+                    end_soc = -0.1
```

Comment:
what? This needs a good really good explanation. What is this trying to do and why in 10% steps?

Reply:
it checks whether the desired soc can be reached at the end of the standing time of the vehicle. If not, the discharge limit of the vehicle battery is increased. It could be smaller steps, what do you suggest?

Reply:
checks whether the desired soc can be reached in the end of the standing time of the vehicle. If not, the minimal soc for discharge is raised. It could be smaller steps, what do you suggest?